app/controllers/my-projects/assignment/recommendations/add.js

```diff
@@ -257,49 +270,83 @@ export default class MyProjectsProjectRecommendationsAddController extends Contr
    */
   @action
   async submitRecommendations() {
-    const { participantType } = this;
-    const targetField = RECOMMENDATION_FIELD_BY_PARTICIPANT_TYPE_LOOKUP[participantType];
-    this.dispositionForAllActionsChangeset.execute();
-    this.dispositionsChangesets.forEach(function(dispositionChangeset) {
-      dispositionChangeset.execute();
-    });
-    // statuscode = "Submitted" signifies in database that user has submitted recommendation
-    this.dispositions.setEach('statuscode', 'Submitted');
+    // array of true/false values each representing succesful upload of files to a disposition
+    const uploadResults = [];
+
+    try {
+      // loop through dispositions, uploading all files to each one
+      for (let i = 0; i < this.dispositions.length; i += 1) {
+        const disposition = this.dispositions.objectAt(i);
+
+        const fileUploadPromises = this.recommendationAddQueue.files.map(file => file.upload(`${ENV.host}/document`, {
+          fileKey: 'file',
+          data: {
+            instanceId: disposition.id,
+            entityName: 'dcp_communityboarddisposition',
+          },
+        }));
+
+        // await here because ember-file-upload requires a single file's upload() to complete before the
+        // next call of upload() on that file.
+        const fileUploadResponses = await Promise.all(fileUploadPromises); // eslint-disable-line
+
+        const filesUploadedToDispo = fileUploadResponses.every(res => res.status === 200);
 
-    this.dispositions.forEach((disposition) => {
-      if (this.allActions) {
-        disposition.set(targetField, this.dispositionForAllActions.recommendation);
+        uploadResults.push(filesUploadedToDispo);
+      }
```

Comment:
I really don't think we should use for loops. The reason is that turns everything synchronous. So each await that gets triggered in the loop waits for the last one. See this: https://zellwk.com/blog/async-await-in-loops/

This can be achieved with array maps, reducers, etc.

Reply:
Great article, thank you for the share 🙇

Reply:
Alright, but I would say that this is still possible _without_ a for loop, and that if you needed to control the concurrency, you can do that through other more expressive means. We can still merge this but I'm personally not convinced there's no way to avoid a synchronous for-loop.